spec/models/facility_spec.rb

```diff
@@ -53,20 +53,15 @@
     it { should delegate_method(:follow_ups_by_period).to(:patients).with_prefix(:patient) }
 
     describe ".assigned_hypertension_patients" do
-      let!(:assigned_facility) { create(:facility) }
-      let!(:registration_facility) { create(:facility) }
-      let!(:assigned_patients) do
-        [create(:patient,
-          assigned_facility: assigned_facility,
-          registration_facility: registration_facility),
-          create(:patient,
-            :without_hypertension,
-            assigned_facility: assigned_facility,
-            registration_facility: registration_facility)]
+      let(:assigned_facility) { create(:facility) }
+      let(:registration_facility) { create(:facility) }
+      before do
+        @assigned_patients = [create(:patient, assigned_facility: assigned_facility, registration_facility: registration_facility),
+          create(:patient, :without_hypertension, assigned_facility: assigned_facility, registration_facility: registration_facility)]
       end
```

Comment:
I feel that creating an instance var in a `before` is just another way of writing a `let!`. Moreover, instance variables are hard to follow for. I've avoided them: 
- Instance variables don't throw a `NameError` when they're not declared, they just silently evaluate to `nil`
- It's possible to share state across specs in [funky ways](https://tomdebruijn.com/posts/ruby-rspec-instance-variables/#modifying-instance-variable-values).

Unless there is a strong reason to move to using instance vars, we should stick with plain old `let!`s. I think when we scrap a `let!` away, the only places that code should move to are:
- inside the `it` block
- into a method that is called in the `it` block
- into a factory method that is called in the `it` block

imo these other Rspec helpers share the same problem that `let!` has and are worse in their own ways.

Reply:
For context here, I refactored from a mess of `let!` to a slightly less messy solution which is at least more clear to follow involving the `before` created state. Some more details below:

> I feel that creating an instance var in a `before` is just another way of writing a `let!`. Moreover, instance variables are harder to follow:

> * Instance variables don't throw a `NameError` when they're not declared, they just silently evaluate to `nil`

Agreed, that is not good. 

> * It's possible to share state across specs in [funky ways](https://tomdebruijn.com/posts/ruby-rspec-instance-variables/#modifying-instance-variable-values).

That post is talking about `before(:context)`, which is not the same as a plain `before`.  `Before(:context)` is setup once per context and shared with all specs inside the context, whereas standard `before` is reset and run for every spec. Same as any regular test state. Note that a `before` with no args is the same as `before(:example)` - see docs here: https://github.com/rspec/rspec-core/blob/c6315d6e899796d3d0203dc8b656708a3ebca9a1/lib/rspec/core/hooks.rb#L14-L15. The different args to `before` and the default arg at play can make this confusing 😐.

I agree that other `before` setup _besides_ `example` should be used very minimally, I've only added them for some [very isolated cases](https://github.com/simpledotorg/simple-server/blob/5a21196f7db598389b12e539d27ca63761c163f3/spec/rails_helper.rb#L31-L34).

> Unless there is a strong reason to move to using instance vars, we should stick with plain old `let!`s. I think when we scrap a `let!`, the only places that code should move to are:

> * inside the `it` block
> * into a method that is called in the `it` block

I agree, both of the above options are often preferable to me. They take more time to refactor to though when you have a lot specs using similar test setup (like this spec did).

> * into a factory method that is called in the `it` block

I'd like to avoid more logic in our factories if possible, unless its really something that is widely shared across our test suite.

> imo these other Rspec helpers share the same problem that `let!` has and are worse in their own ways.

I agree that ivars and `attr_reader` in tests can look weird. For context, I was working to remove our reliance on `let!` as well as to improve our test speed. I agree that there could be more done here to refactor test setup to plain locals or regular ruby methods to make things more clear.  For this specific case, maybe a regular ruby method that creates the assigned patients would be more clear?

Reply:
Yeah I've relied on regular ruby methods since we started moving away from `let`. 

This looks good for a cleanup. I guess what I don't follow here is the `@assigned_patients` in `before`. It could also be a `let` like the other ones, no?

PS It'd be great if we can also follow something like "No instance vars in our tests" 😛  . wdyt?

Reply:
> This looks good for a cleanup. I guess what I don't follow here is the `@assigned_patients` in `before`. It could also be a `let` like the other ones, no?

Yeah I agree the ivar also looks weird, it is a tradeoff between more deterministic setup and the ease of 'cleanliness' of a let block.  I could go either way on that.